src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndexForAdd(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null; // Clear the last element
```

Comment:
```suggestion
        elements[--size] = null;
```
don't forget to remove your comments